samza-core/src/main/java/org/apache/samza/container/grouper/task/TaskNameGrouper.java

```diff
@@ -44,15 +43,36 @@
  * </p>
  */
 public interface TaskNameGrouper {
+
+  /**
+   * Groups the taskModels into set of {@link ContainerModel} using the provided grouperContext.
+   *
+   * @param taskModels the set of tasks to group into containers.
+   * @param grouperContext the existing historical metadata of the samza job.
+   * @return the grouped {@link ContainerModel} built from the provided taskModels.
+   */
+  default Set<ContainerModel> group(Set<TaskModel> taskModels, GrouperContext grouperContext) {
```

Comment:
Why is the GrouperContext part of the group method? I feel like the context is more of a metadata that the implementor needs in order to perform grouping. It may well be useful beyond the grouping path like some setup before performing grouping.

Although I don't like the idea of having lifecycle methods, can we have an init method that takes the context and keep the group method clean? Are there any implications that requires the current pattern?

Ideally we can pass in the context during construction though I am not sure if it is going to complicate things here since we use reflection.